dht_rt_seeder.go

```diff
@@ -0,0 +1,119 @@
+package dht
+
+import (
+	"context"
+	"time"
+
+	"github.com/ipfs/go-todocounter"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// SeederDialTimeout is the grace period for one dial attempt
+var SeederDialTimeout = 5 * time.Second
+
+// TotalSeederDialTimeout is the total grace period for a group of dial attempts
+var TotalSeederDialTimeout = 30 * time.Second
+
+// SeederConcurrentDials is the number of peers we will dial simultaneously
+var SeederConcurrentDials = 50
+
+func (dht *IpfsDHT) seedRoutingTable(candidates, fallbacks []peer.ID) error {
+	peersFromSet := func(m map[peer.ID]struct{}) []peer.ID {
+		var peers []peer.ID
+		for p := range m {
+			peers = append(peers, p)
+		}
+
+		return peers
+	}
+
+	// candidates we have still not tried dialling to
+	remainingCandidates := make(map[peer.ID]struct{})
+	// fallbacks we have still not tried dialling to
+	remainingFallbacks := make(map[peer.ID]struct{})
+
+	for i := range candidates {
+		remainingCandidates[candidates[i]] = struct{}{}
+	}
+
+	for i := range fallbacks {
+		remainingFallbacks[fallbacks[i]] = struct{}{}
+	}
+
+	for {
+
+		todial := dht.seedsProposer.Propose(dht.routingTable, peersFromSet(remainingCandidates), peersFromSet(remainingFallbacks))
```

Comment:
Taking into account https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r357988342.

I'd like to understand a little better what we're trying to do here. The logic here seems to work fine, but depending on what we're trying to accomplish it might be overly complicated or a little confusing. I'm not recommending any blocking code changes here, but just trying to understand/clarify some of the interface considerations.

1) We propose peers based on how good the peers are (no routing table info)
   - How good peers are probably won't change much during a seeding, so no repeated computation is really required here
   - Why not just return a channel of peers and keep pulling off peers until we've hit the number of seeders we care about?
2) We propose peers based on how good the peers are **AND** the state of the routing table
   - We potentially have two operations that we need to run repeatedly: Propose and Dial
   - If Propose is more expensive then Dial then the current approach works well. We Propose a bunch of peers and Dial some subset (`seederConcurrentDials`) of them at once. Once we've Dialed all the proposed peers we call Propose again and restart this procedure.
        - However, we're not necessarily dialing peers in the optimal order. For example, we really want to fill bucket X with one peer, but that peer dial fails, it would be optimal to try the next peer in that bucket as the next dial but we won't because of the batching
   - If Dial is much more expensive then Propose (I expect this is the case), then it'd be nice if we could initially propose `seederConcurrentDials` peers and then Propose a single new peer whenever a goroutine becomes free. This way we're always dialing the most valuable next peer.
      - We could have something close to this currently by just suggesting state-based implementations of Propose only return a one (or a few) peers at a time. A little more expensive for the first set of dials, but probably not a big deal.
      - If Dial is not much more expensive then Propose then this is wasteful

@aarshkshah1992 am I describing the situation accurately? Does my recommendation for state-based Proposers to return very small sets of peers make sense, or am I misunderstanding how this is supposed to work?

Reply:
@aschmahmann  I think you have the nail on the head. Dial can be assumed to be more expensive than propose.

For now, I have made the _Proposer_ return a channel which it will close when it's done proposing. This greatly simplifies the code. 

Please can you elaborate a bit on the advantage of having the proposer returning small peer sets rather than the the whole ordering on the channel like we do in this new implementation ?

Reply:
> Is it because the repeated interplay between dial & propose makes it easy for the proposer to observer RT state changes across dials ?

Yes. This is why I asked if we cared about RT state, or just peer quality. If we only care about peer quality (or just use the random proposer) then using a channel is fine. However, the proposer cannot submit the "optimal peer" unless it has the latest routing table information.

While we may still be able to use channels, if we really are just submitting proposed peers one (or a few) at a time we may as well do that with a function call.